### 1.01: Help and Documentation

Doing ? after something in Ipython gives info on it, and ?? can be helpful to access source code

Doing L. and then tab will give options on object contents



### 1.02: Shell Keyboard Shortcuts

| Keystroke                         | Action                                     |
|-----------------------------------|--------------------------------------------|
| ``Ctrl-a``                        | Move cursor to the beginning of the line   |
| ``Ctrl-e``                        | Move cursor to the end of the line         |
| ``Ctrl-b`` or the left arrow key  | Move cursor back one character             |
| ``Ctrl-f`` or the right arrow key | Move cursor forward one character          |
| Backspace key                 | Delete previous character in line                |
| ``Ctrl-d``                    | Delete next character in line                    |
| ``Ctrl-k``                    | Cut text from cursor to end of line              |
| ``Ctrl-u``                    | Cut text from beginning of line to cursor        |
| ``Ctrl-y``                    | Yank (i.e. paste) text that was previously cut   |
| ``Ctrl-t``                    | Transpose (i.e., switch) previous two characters |
| ``Ctrl-p`` (or the up arrow key)    | Access previous command in history         |
| ``Ctrl-n`` (or the down arrow key)  | Access next command in history             |
| ``Ctrl-r``                          | Reverse-search through command history     |
| ``Ctrl-l``                    | Clear terminal screen                      |
| ``Ctrl-c``                    | Interrupt current Python command           |
| ``Ctrl-d``                    | Exit IPython session                       |



### 1.03: Magic Commands

**Magic commands** are a feature in IPython added on top of the normal python syntax.
- **line magic**: denoted by a single % prefix and operate on a single line of output
- **cell magic**: deonted by a double %% prefix and operate on multiple lines of input

``%paste`` is designed to handle multi-line, marked-up input
ex: 
```ipython
In [3]: %paste
>>> def donothing(x):
...     return x

```

The ``%paste%`` command enters and executes the code, so the function is ready to be used in the interpreter. ``%cpaste`` is similar, but it allows you to paste more than one chunk of code to be executed in a batch.


#### Running external code: ``%run``

Pretty much just runs an external script.

``%timeit`` will determine the execution time of the single-line Pythong statement that follows it. 

``%magic`` will list all available magic functions



### 1.4: Input and Output History

Inputs and outputs can give a clue as to how you can access previous inputs and ouptuts in your current session. It allows you to access previous information that was inputted our outputted by calling, for example, ``print(Out[2])``.



### 1.05: IPython and Shell Commands

Use ! on a line to execute by system command-line (shell)

The shell is a way to interact textually with your computer. 

ex:

In [1]:
!ls
!pwd

__pycache__             helpful-stuff_ch1.ipynb
helpful-stuff.ipynb     mprun_demo.py
/Users/maxwellpatterson/Desktop/personal/PythonDataScienceHandbook/notebooks_v1/xwell-notes


Shell commands can not only be called from IPython, but cna also be made to interact with the IPython namespace. For ex, you can save the outut of any shell command to a Python list using the assignment operator. Helpful!

### 1.6: Errors and Debugging

With the ``%xmode`` magic function, IPython allows you to control the amout of information printed when the exception is raised.

Let's see an example here:

In [2]:
def func1(a, b):
    return a / b

def func2(x):
    a = x
    b = x - 1
    return func1(a, b)

func2(1)

ZeroDivisionError: division by zero

Obviously this will throw an error since we are dividing by zero.

``%xmode`` can be one of three options: Plain, Context, and Verbose. Context is the default and gives less information.

In [ ]:
%xmode Plain

func2(1)

%xmode Verbose
func2(1)

Exception reporting mode: Plain


ZeroDivisionError: division by zero

This extra info from Verbose can be helpful. However, when working with lots of code and complicated errors, Default mode can be the best.

Adding ``%pdb on`` to code will launch the debugger automatically whenever an exception is raised.

### 1.7: Profiling and Timing Code

Efficiency is important in coding. IPython provides access to a wide array of funcionality for timing and profiling code. Here are a few magic commands:
- ``%time``: time the execution of a single statement
- ``%timeit``: time repeated execution of a single statement for more accuracty
- ``%prun``: run code with the profiler
- ``%lprun``: run code with the line-by-line profiler
- ``%memit``: measure the memory use of a single statement
- ``%mprun``: run code with the line-by-line memory profiler

Note that the last four will need ``line_profiler`` and ``memory_profiler`` extensions.

Let's see some examples of how ``%timeit`` works:

In [ ]:
%timeit sum(range(100))

658 ns ± 2.25 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [ ]:
%%timeit
total = 0
for i in range(1000):
    for j in range(1000):
        total += i * (-1) ** j

248 ms ± 1.92 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


Now let's move on to profiling full scripts with ``%prun``. A program is made of a bunch of single statements, and sometimes timing these statements in context is more important than timing them on their own. Let's see an example:

In [ ]:
def sum_of_lists(N):
    total = 0
    for i in range(5):
        L = [j ^ (j >> i) for j in range(N)]
        total += sum(L)
    return total

In [ ]:
%prun sum_of_lists(1000000)

         14 function calls in 0.480 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        5    0.417    0.083    0.417    0.083 <ipython-input-3-f105717832a2>:4(<listcomp>)
        5    0.030    0.006    0.030    0.006 {built-in method builtins.sum}
        1    0.028    0.028    0.474    0.474 <ipython-input-3-f105717832a2>:1(sum_of_lists)
        1    0.006    0.006    0.480    0.480 <string>:1(<module>)
        1    0.000    0.000    0.480    0.480 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}

The table outputted above indicates, in order f total time on each call, where the execution is spending the most time. In this case, the bulk of execution time is from ``sum_of_lists``. 

How about we go line-by-line with ``%lprun``. First, we need to actually install it.

In [ ]:
!pip install line_profiler
%load_ext line_profiler

Great. Now, the ``%lprun`` command will do a line-by-line profiling of any function. Let's see an example using the sum_of_lists function:

In [ ]:
%lprun -f sum_of_lists sum_of_lists(5000)

Timer unit: 1e-09 s

Total time: 0.014134 s
File: <ipython-input-3-f105717832a2>
Function: sum_of_lists at line 1

Line #      Hits         Time  Per Hit   % Time  Line Contents
     1                                           def sum_of_lists(N):
     2         1       1000.0   1000.0      0.0      total = 0
     3         6      15000.0   2500.0      0.1      for i in range(5):
     4         5   13861000.0    3e+06     98.1          L = [j ^ (j >> i) for j in range(N)]
     5         5     257000.0  51400.0      1.8          total += sum(L)
     6         1          0.0      0.0      0.0      return total

The info at the top gives us the key to interpreting results: the time is reported in microseconds and we can see where the program is spending the most time. We can use this info to modify aspects of the scripts and make it perform better for our desired use case. 

Another aspect of profiling is the amount of memory an operation uses. This can be evaluated with the ``memory_profiler`` extension. Let's install that

In [ ]:
!pip install memory_profiler

In [ ]:
# Load memory_profiler extension
%load_ext memory_profiler

The memory profiler extension contains two useful magig functions: 
- ``%memit``: offers a memory-measuring equivalent of ``%timeint``
- ``%mprun``: offers a memory-measuring equivalent of ``%lprun``

In [ ]:
%memit sum_of_lists(1000000)

peak memory: 231.22 MiB, increment: 63.23 MiB


The output tells us that this function uses about 231 MB of memory (see peak memory). 

``%mprun`` gives us a line-by-line description of memory use. Unfortunately, this magic works only for functions defined in separate modules rather than the notebook itself, so we will starty by using the ``%%file`` magic to create a simple module called ``mprun_demo.py``, which contains the ``sum_of_lists`` function, with one addition that will make our memory profiling results clearer:

In [ ]:
%%file mprun_demo.py
def sum_of_lists(N):
    total = 0
    for i in range(5):
        L = [j ^ (j >> i) for j in range(N)]
        total += sum(L)
        del L # remove reference to L
    return total

Writing mprun_demo.py


In [ ]:
from mprun_demo import sum_of_lists
%mprun -f sum_of_lists sum_of_lists(1000000)

*** KeyboardInterrupt exception caught in code being profiled.


Filename: /Users/maxwellpatterson/Desktop/personal/PythonDataScienceHandbook/notebooks_v1/xwell-notes/mprun_demo.py

Line #    Mem usage    Increment  Occurrences   Line Contents
     1    232.6 MiB    232.6 MiB           1   def sum_of_lists(N):
     2    232.6 MiB      0.0 MiB           1       total = 0
     3    232.6 MiB    -13.2 MiB           4       for i in range(5):
     4    232.6 MiB -45847082.7 MiB     3567554           L = [j ^ (j >> i) for j in range(N)]
     5    230.3 MiB    -20.1 MiB           3           total += sum(L)
     6    222.7 MiB    -36.1 MiB           3           del L # remove reference to L
     7                                             return total

The ``Increment`` column tells us how much each line affects the total memory budget.